### DISCUSSION TITLE
Poor Gradient Validation for Streamwise Periodic Flow solver with type MASSFLOW.

### DISCUSSION THREAD

**Discussion by NAnand-TUD** [su2code/SU2#1709] [General]
Hello! 

I am testing the streamwise periodic (SP) adjoint solver and found that the Gradient Validation is spot on for PRESSURE_DROP based SP solver, however, with MASSFLOW based SP solver the gradients are off by more than twice the value. @TobiKattmann any clue why that might be the case ?

I am currently implementing an additional term in SP solver which needs to be updated as pressure drop is being updated in MASSFLOW SP.

Thanks,
Nitish

**Comment by TobiKattmann** [comment]
Are you missing code from this PR by any chance #1536 (merged into develop on Feb 26 and part of 7.3.1. master version)

There we added an 'additional adjoint adjoint equation' for pressure-drop as it is also a fixed point equation just as the flow solution and therefore needs to be handled alike.

**Comment by NAnand-TUD** [comment]
Yes, I am up to date with the develop branch so I am certain that I am including that the new adjoint equation is being accounted for. I see that there is AD::Register for PressureDrop in the version I am using. 

To be double sure I also ran on the commit which includes your updates from 26th Feb, however, still no validation for the MassFlow-based SP solver. Can you kindly share your config file or point me to the one in the TestCases that you used for validation? Thanks!

**Reply by TobiKattmann** [reply]
I added `incomp_navierstokes/streamwise_periodic/dp-adjoint_chtPinArray_2d` in that very PR. Not sure why I didn't add a FADO gradient validation script... but you seem to have a setup for FD yourself.

Another thing: Does the discrete adjoint run has access to the `flow.meta` file which is written by the primal? That contains the pressure drop value for a good restart. (not sure whether I decided to throw an error then, just a warning or nothing :thinking: )

**Reply by TobiKattmann** [reply]
A warning is written to screen output during solver preprocessing `Warning: There is no restart file (flow_0.meta)`

**Reply by NAnand-TUD** [reply]
Yes, I have FD infrastructure inside ParaBlade to do that. Correct, so the *.meta file is being written and read correctly. Additionally, the residual readout of the primal solver during the adjoint computation is correct. 

Do you see anything out of place in the config file? 

`%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                              %
% SU2 configuration file                                                       %
% Case description: Steady incompressible laminar flow around heated cylinders %
%                                                                              %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% ------------- DIRECT, ADJOINT, AND LINEARIZED PROBLEM DEFINITION ------------%
%
% Physical governing equations (EULER, NAVIER_STOKES,
%                               WAVE_EQUATION, HEAT_EQUATION, FEM_ELASTICITY,
%                               POISSON_EQUATION)                           
SOLVER= INC_NAVIER_STOKES
%
% If Navier-Stokes, kind of turbulent model (NONE, SA)
KIND_TURB_MODEL= NONE
%
% Restart solution (NO, YES)
RESTART_SOL= NO
%
% Objective function in gradient evaluation   (DRAG, LIFT, SIDEFORCE, MOMENT_X,
%                                             MOMENT_Y, MOMENT_Z, EFFICIENCY,
%                                             EQUIVALENT_AREA, NEARFIELD_PRESSURE,
%                                             FORCE_X, FORCE_Y, FORCE_Z, THRUST,
%                                             TORQUE, TOTAL_HEATFLUX,
%                                             MAXIMUM_HEATFLUX, INVERSE_DESIGN_PRESSURE,
%                                             INVERSE_DESIGN_HEATFLUX, SURFACE_TOTAL_PRESSURE,
%                                             SURFACE_MASSFLOW, SURFACE_STATIC_PRESSURE, SURFACE_MACH
% For a weighted sum of objectives: separate by commas, add OBJECTIVE_WEIGHT and MARKER_MONITORING in matching order.
OBJECTIVE_FUNCTION = TOTAL_HEATFLUX
%
% List of weighting values when using more than one OBJECTIVE_FUNCTION. Separate by commas and match with MARKER_MONITORING.
OBJECTIVE_WEIGHT = 1.0
%
% Read binary restart files (YES, NO)
READ_BINARY_RESTART = YES
%
% Data written to history file
HISTORY_OUTPUT=(ITER, RMS_RES, HEAT, FLOW_COEFF, AERO_COEFF, STREAMWISE_PERIODIC)

% -------------------- BOUNDARY CONDITION DEFINITION --------------------------%
%
% Farfield boundary marker(s) (NONE = no marker)
%MARKER_FAR= ( outmix )
%
MARKER_SYM= (periodic1, periodic2)
%MARKER_ISOTHERMAL= (Blade, 0.0)
MARKER_HEATFLUX= (Blade, 100.0)
% Marker(s) of the surface to be plotted or designed
MARKER_PLOTTING= (Blade)
%
% Marker(s) of the surface where the functional (Cd, Cl, etc.) will be evaluated
MARKER_MONITORING= (Blade)
MARKER_ANALYZE= (Blade)
%INC_INLET_TYPE= VELOCITY_INLET
%INC_OUTLET_TYPE= PRESSURE_OUTLET
%
%MARKER_HEATFLUX= ( Blade, 0.0)
%
%MARKER_PERIODIC=   ( periodic1, periodic2, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -2.0, 0.0)
%
%MARKER_INLET= ( inflow, 288.15, 0.55, 1.0, 0.0, 0.0 )
%
%MARKER_OUTLET= ( outflow, 0.0 ) 

% --------------------- STREAMWISE PERIODICITY DEFINITION ---------------------%
%
KIND_STREAMWISE_PERIODIC= MASSFLOW
STREAMWISE_PERIODIC_MASSFLOW= 0.326
% massflow = rho * A * V = 1.28* 0.061 * 3.6 = 0.281
STREAMWISE_PERIODIC_PRESSURE_DROP= 0.03
INC_OUTLET_DAMPING= 0.01
%
MARKER_PERIODIC= ( inflow, outflow, 0.0,0.0,0.0, 0.0,0.0,0.0, 2.0,0.0,0.0) %, periodic1, periodic2, 0.0,0.0,0.0, 0.0,0.0,0.0, 0.0,-2.0,0.0) 
STREAMWISE_PERIODIC_TEMPERATURE= YES
% Computation of outlet heat: Heatflux * Area 
%                             280[W/m] * 19 * 0.00635[m]
%STREAMWISE_PERIODIC_OUTLET_HEAT= -33.782 

% ---------------- INCOMPRESSIBLE FLOW CONDITION DEFINITION -------------------%
%
% Density model within the incompressible flow solver.
% Options are CONSTANT (default), BOUSSINESQ, or VARIABLE. If VARIABLE,
% an appropriate fluid model must be selected.
INC_DENSITY_MODEL= CONSTANT
%
% Solve the energy equation in the incompressible flow solver
INC_ENERGY_EQUATION = YES
%
% Initial density for incompressible flows (1.2886 kg/m^3 by default)
INC_DENSITY_INIT= 1.2886
%
% Initial velocity for incompressible flows (1.0,0,0 m/s by default)
INC_VELOCITY_INIT= ( 0.1, 0.0, 0.0 )
%
% Initial temperature for incompressible flows that include the
% energy equation (288.15 K by default). Value is ignored if
% INC_ENERGY_EQUATION is false.
INC_TEMPERATURE_INIT= 288.15
%
% Non-dimensionalization scheme for incompressible flows. Options are
% INITIAL_VALUES (default), REFERENCE_VALUES, or DIMENSIONAL.
% INC_*_REF values are ignored unless REFERENCE_VALUES is chosen.
INC_NONDIM= DIMENSIONAL

% ---- IDEAL GAS, POLYTROPIC, VAN DER WAALS AND PENG ROBINSON CONSTANTS -------%
%
% Fluid model (STANDARD_AIR, IDEAL_GAS, VW_GAS, PR_GAS,
%              CONSTANT_DENSITY, INC_IDEAL_GAS)
FLUID_MODEL= INC_IDEAL_GAS
%
% Specific heat at constant pressure, Cp (1004.703 J/kg*K (air)).
% Incompressible fluids with energy eqn. only (CONSTANT_DENSITY, INC_IDEAL_GAS).
SPECIFIC_HEAT_CP= 1004.703
%
% Molecular weight for an incompressible ideal gas (28.96 g/mol (air) default)
% Incompressible fluids with energy eqn. only (CONSTANT_DENSITY, INC_IDEAL_GAS).
MOLECULAR_WEIGHT= 28.96

% --------------------------- VISCOSITY MODEL ---------------------------------%
%
% Viscosity model (SUTHERLAND, CONSTANT_VISCOSITY).
VISCOSITY_MODEL= CONSTANT_VISCOSITY
%
% Molecular Viscosity that would be constant (1.716E-5 by default)
MU_CONSTANT= 0.005558784
%
% Sutherland Viscosity Ref (1.716E-5 default value for AIR SI)
MU_REF= 1.716E-5
%
% Sutherland Temperature Ref (273.15 K default value for AIR SI)
MU_T_REF= 273.15
%
% Sutherland constant (110.4 default value for AIR SI)
SUTHERLAND_CONSTANT= 110.4

% --------------------------- THERMAL CONDUCTIVITY MODEL ----------------------%
%
% Conductivity model (CONSTANT_CONDUCTIVITY, CONSTANT_PRANDTL).
CONDUCTIVITY_MODEL= CONSTANT_PRANDTL
%
% Molecular Thermal Conductivity that would be constant (0.0257 by default)
THERMAL_CONDUCTIVITY_CONSTANT= 0.0257
%
% Laminar Prandtl number (0.72 (air), only for CONSTANT_PRANDTL)
PRANDTL_LAM= 0.72
%
% Turbulent Prandtl number (0.9 (air), only for CONSTANT_PRANDTL)
PRANDTL_TURB= 0.90

% ------------- COMMON PARAMETERS DEFINING THE NUMERICAL METHOD ---------------%
%
% Numerical method for spatial gradients (GREEN_GAUSS, WEIGHTED_LEAST_SQUARES)
NUM_METHOD_GRAD= GREEN_GAUSS
%
% Courant-Friedrichs-Lewy condition of the finest grid
CFL_NUMBER= 1e3
%
% Adaptive CFL number (NO, YES)
CFL_ADAPT= YES
%
% Parameters of the adaptive CFL number (factor down, factor up, CFL min value,
%                                        CFL max value )
CFL_ADAPT_PARAM= ( 0.4, 1.5, 5.0, 10000.0 )
%
% Runge-Kutta alpha coefficients
RK_ALPHA_COEFF= ( 0.66667, 0.66667, 1.000000 )

% ------------------------ LINEAR SOLVER DEFINITION ---------------------------%
%
% Linear solver or smoother for implicit formulations (BCGSTAB, FGMRES, SMOOTHER_JACOBI,
%                                                      SMOOTHER_ILU, SMOOTHER_LUSGS,
%                                                      SMOOTHER_LINELET)
LINEAR_SOLVER= FGMRES
%
% Preconditioner of the Krylov linear solver (ILU, LU_SGS, LINELET, JACOBI)
LINEAR_SOLVER_PREC= ILU
%
% Linear solver ILU preconditioner fill-in level (0 by default)
LINEAR_SOLVER_ILU_FILL_IN= 0
%
% Minimum error of the linear solver for implicit formulations
LINEAR_SOLVER_ERROR= 1E-15
%
% Max number of iterations of the linear solver for the implicit formulation
LINEAR_SOLVER_ITER= 10

% -------------------- FLOW NUMERICAL METHOD DEFINITION -----------------------%
%
% Convective numerical method (JST, LAX-FRIEDRICH, CUSP, ROE, AUSM, HLLC,
%                              TURKEL_PREC, MSW)
CONV_NUM_METHOD_FLOW= FDS
%
% Monotonic Upwind Scheme for Conservation Laws (TVD) in the flow equations.
%           Required for 2nd order upwind schemes (NO, YES)
MUSCL_FLOW= NO
%
% Slope limiter (NONE, VENKATAKRISHNAN, VENKATAKRISHNAN_WANG,
%                BARTH_JESPERSEN, VAN_ALBADA_EDGE)
SLOPE_LIMITER_FLOW= NONE
%
% Time discretization (RUNGE-KUTTA_EXPLICIT, EULER_IMPLICIT, EULER_EXPLICIT)
TIME_DISCRE_FLOW= EULER_IMPLICIT

% --------------------------- CONVERGENCE PARAMETERS --------------------------%
%
% Min value of the residual (log10 of the residual)
CONV_RESIDUAL_MINVAL= -9
%
% Start convergence criteria at iteration number
CONV_STARTITER= 4
%
CONV_FIELD= RES_PRESSURE
%
% Number of elements to apply the criteria
CONV_CAUCHY_ELEMS= 100
%
% Epsilon to control the series convergence
CONV_CAUCHY_EPS= 1E-6

% ------------------------- INPUT/OUTPUT INFORMATION --------------------------%
%
% Restart flow input file
SOLUTION_FILENAME= restart_flow.dat
%
% Restart adjoint input file
SOLUTION_ADJ_FILENAME= restart_adj.dat
%
TABULAR_FORMAT= TECPLOT
%
% Output file convergence history (w/o extension)
CONV_FILENAME= history
%
% Output file with the forces breakdown
BREAKDOWN_FILENAME= forces_breakdown.dat
%
% Output file restart flow
RESTART_FILENAME= restart_flow.dat
%
% Output file restart adjoint
RESTART_ADJ_FILENAME= restart_adj.dat
%
% Output file flow (w/o extension) variables
VOLUME_FILENAME= flow
%
% Output file adjoint (w/o extension) variables
VOLUME_ADJ_FILENAME= adjoint
%
% Output Objective function
VALUE_OBJFUNC_FILENAME= of_eval.dat
%
% Output objective function gradient (using continuous adjoint)
GRAD_OBJFUNC_FILENAME= of_grad.dat
%
% Output file surface flow coefficient (w/o extension)
SURFACE_FILENAME= surface_flow
%
% Output file surface adjoint coefficient (w/o extension)
SURFACE_ADJ_FILENAME= surface_adjoint
%
OUTPUT_FILES= (PARAVIEW_ASCII, SURFACE_TECPLOT_ASCII, RESTART)
% ------------------------ GRID DEFORMATION PARAMETERS ------------------------%
%
% Linear solver or smoother for implicit formulations (FGMRES, RESTARTED_FGMRES, BCGSTAB)
DEFORM_LINEAR_SOLVER= FGMRES
%
% Number of smoothing iterations for mesh deformation
DEFORM_LINEAR_SOLVER_ITER= 200
%
% Number of nonlinear deformation iterations (surface deformation increments)
DEFORM_NONLINEAR_ITER= 1
%
% Print the residuals during mesh deformation to the console (YES, NO)
DEFORM_CONSOLE_OUTPUT= YES
%
% Type of element stiffness imposed for FEA mesh deformation (INVERSE_VOLUME,
%                                          WALL_DISTANCE, CONSTANT_STIFFNESS)
DEFORM_STIFFNESS_TYPE= INVERSE_VOLUME

INNER_ITER= 6000

% ------------------------ MESH ------------------------%
%
% Mesh input file
MESH_FILENAME= rotor_domain.su2
MESH_OUT_FILENAME= rotor_domain_out.su2
%
% Mesh input file format (SU2, CGNS, NETCDF_ASCII)
MESH_FORMAT= SU2

%-------------------------- DESIGN -------------------%
DV_KIND= SURFACE_FILE
DV_MARKER= (Blade)
DV_FILENAME= MoveSurface.txt 
GEO_MARKER= (Blade)
OUTPUT_WRT_FREQ= 500
DV_PARAM=  ( MAIN_BOX, 12, 1, 0.0, 1.0 )
DV_VALUE=  0.1
`

**Reply by TobiKattmann** [reply]
Hm, your OF is `TOTAL_HEATFLUX` (Area integral of Heatflux) on `Blade`... which is a constant `MARKER_HEATFLUX`, i.e. I expect the sensitivities to be perpendicular to the `Blade` surface (outward facing) -> That expands the surface the most and thus creates the maximum increase to the OF. And that should be totally independent of the flow around it.

Does that make sense, or do I have a mistake in my thought 

The sensitivities should be computed correctly though. (Edit: which they seemingly are not ;) )

**Reply by TobiKattmann** [reply]
(Tiny complaint incoming :flying_saucer: ) And it definitely helps to provide cfg's/cases in a bit more cleaned-up state. I had to remove all unnecessary options and  regroup a few things before I could see things

[tidy.txt](https://github.com/su2code/SU2/files/9113077/tidy.txt)

Can you maybe try 
`OBJECTIVE_FUNCTION= SURFACE_STATIC_TEMPERATURE` (`Avg_Temp` in the history) with `MARKER_ANALYZE= Blade` and `MARKER_ANALYZE_AVERAGE=AREA` as well as setting `STREAMWISE_PERIODIC_TEMPERATURE= NO` (then the inlet temp will we averaged to `INC_INLET_TEMPERATURE`)

That setup is a bit closer to what I used a lot -> (this is of course a try to circle the issue)

I assume you checked that primal and adjoint converge as well as you are using the correct values from the history to compute the FD to the OF ;)

**Reply by NAnand-TUD** [reply]
Hmm, yes but I was still getting Gradient Validation for HeatFlux, maybe because of a change in area? 
Anyhow, after running a ton of tests I can see that the adjoint convergence is poor when the energy equations are turned on. Any recommendation on how to get that improved?

Also, thanks for the suggestion on cleaning the config file. ;)

**Reply by NAnand-TUD** [reply]
@TobiKattmann: So I ran multiple cases and I can see that getting adjoint convergence with SP-Temp is extremely difficult and hence no gradient validation. For cases without Temp Eqn. I am able to get gradient validation. Am pretty lost so as to what is going wrong. Just to make sure, can you please run validation at your end using the Fado framework so as to confirm that there is something fishy in my case itself.

**Reply by NAnand-TUD** [reply]
Maybe if you run a case without Conjugate Heat would allow us to compare apples to apples...

**Reply by TobiKattmann** [reply]
Well, I do have gradient validation using FADO for the testcases I consider. So I do not really see the benefit of re-running those. And I do have a validated non-CHT version of my standard unit-cell pin thing with massflow and energy equation on. If you mean your case, then I do not have the mesh up to now. I can take a look at your case and set up a FADO-GV (most likely not before next Wednesday though).

In the meantime my personal best practice is certainly to use a static CFL (for both direct and adjoint) which you might want to lower for the DA run. Additionally you might want to increase Linear solver iterations for the adjoint. Maybe that already helps as I had some instances where these changes made all the difference between 'zero convergence' and 'same convergence speed as the primal'

But do I understand it correctly now, that MASSFLOW does not seem to the issue but the not-converging adjoint energy equation? (and I might have a blind spot when it comes to the `STREAMWISE_PERIODIC_TEMPERATURE= YES` case, so if you try `NO` while choosing a reasonable 'INC_INLET_TEMPERATURE' that would be of help)

**Reply by NAnand-TUD** [reply]
So here is the conclusion:

Apparently, the TEMP-related objective function was not working very well. Eventually, I implemented an objective function related to LambdaL (which is equivalent to the temperature gradient in SP unit cell) and I started getting validation with pretty good adjoint convergence. 

Here is the validation plot.

[GradVal_h1e-02_HEX_2D.pdf](https://github.com/su2code/SU2/files/9309205/GradVal_h1e-02_HEX_2D.pdf)

I have still not tried validation with a Massflow-based SP solver. I hope I will close that open point as well soon.

**Reply by TobiKattmann** [reply]
Hmm I am wondering whether there is sth funky going on with Temperature based OF when you use streamwise periodic temperature: I would probably make sure, that you use the recovered (physical) temperature value. I did the same for pressure in the respective routine but maybe not for temperature :thinking: 
Because the periodic temperature value only accounts for local effects. Not to sure at the moment whether the discrete adjoint should be consistent in that case as well.

**Reply by NAnand-TUD** [reply]
Hmmm, nice observation. Let me know if you find something.

**Reply by TobiKattmann** [reply]
So in `CFlowOutput.cpp SetAnalyzeSurface`, adding the second line would make it use the recovered (physical) Temperature instead of just the periodic component. The gradient validation works though with and without though. I have a fluid-only case with `SURFACE_STATIC_TEMPERATURE` as an OF and that works fine.
```c++
Temperature       = flow_nodes->GetTemperature(iPoint); // this line already exists
if(streamwisePeriodic) Temperature = flow_nodes->GetStreamwise_Periodic_RecoveredTemperature(iPoint); // this line to add
```

**Reply by NAnand-TUD** [reply]
I see, thanks for checking. I will also check that once. I am scaling the temperature to solve for the iso-thermal wall, maybe that is interfering with the validation.

**Reply by NAnand-TUD** [reply]
@TobiKattmann: I am running some verification for the Turbulent case now, and I find it rather difficult to make a comparison. What test case did you use for the verification of the constant heat-flux problem? 

PS: By verification: I am comparing a 30 in-line circular fin simulation with that of the SWP solution.

**Reply by TobiKattmann** [reply]
I used my standard 2D unit cell setup. It is of course easiest to use a constant massflow in the periodic setup, and then either also constant massflow for the non-periodic simulation but I went for a block-profile with a velocity inlet where of course with fixed density and area the prescribed velocity regulates the domain's massflow. Otherwise there was really nothing special to it. I simply used the downstream profiles at the ''same" periodic location as you see below and that worked well enough for me:
![image](https://user-images.githubusercontent.com/31306376/194755253-d05c87ff-b082-4a58-bd55-6afc8cd43c8e.png)

I am not sure whether I get you concern correctly though

Edit: and for pressure and temperature i simply zero-ed one location for each profile such that one only compares the profiles

**Reply by NAnand-TUD** [reply]
Yes, you got my question right. Periodic BCs were to blame there... Once I switched to SYM BCs the flow converged rather effortlessly... I am still trying to get verification for temperature equations. I will post the plots when I am done implementing them. Plots loading soon....

**Reply by NAnand-TUD** [reply]
Great ! All sorted. Here are my results for iso-thermal BCs...

![image](https://user-images.githubusercontent.com/22835014/195123131-cbdfdd09-96a1-4486-b38f-38cd4e1ffb77.png)

**Reply by NAnand-TUD** [reply]
Hello @TobiKattmann,

So far the Massflow based SP solver has struggled to give consistent gradient validation. Can you tell me what Objective Function you used to validate the Gradients? Can you also share your case? 

-Nitish
